In [7]:
import pandas as pd


df = pd.read_csv("/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-two-years.csv")


print("Forma:", df.shape)
df.head()

Forma: (7214, 53)


id                name   first         last compas_screening_date   sex  \
0   1    miguel hernandez  miguel    hernandez            2013-08-14  Male   
1   3         kevon dixon   kevon        dixon            2013-01-27  Male   
2   4            ed philo      ed        philo            2013-04-14  Male   
3   5         marcu brown   marcu        brown            2013-01-13  Male   
4   6  bouthy pierrelouis  bouthy  pierrelouis            2013-03-26  Male   

          dob  age          age_cat              race  ...  v_decile_score  \
0  1947-04-18   69  Greater than 45             Other  ...               1   
1  1982-01-22   34          25 - 45  African-American  ...               1   
2  1991-05-14   24     Less than 25  African-American  ...               3   
3  1993-01-21   23     Less than 25  African-American  ...               6   
4  1973-01-22   43          25 - 45             Other  ...               1   

   v_score_text  v_screening_date  in_custody  out_custody  priors_count.1  \
0           Low        2013-08-14  2014-07-07   2014-07-14               0   
1           Low        2013-01-27  2013-01-26   2013-02-05               0   
2           Low        2013-04-14  2013-06-16   2013-06-16               4   
3        Medium        2013-01-13         NaN          NaN               1   
4           Low        2013-03-26         NaN          NaN               2   

  start   end event two_year_recid  
0     0   327     0              0  
1     9   159     1              1  
2     0    63     0              1  
3     0  1174     0              0  
4     0  1102     0              0  

[5 rows x 53 columns]

In [8]:
# Tipos de datos
df.dtypes

# Nulos por columna
df.isnull().sum().sort_values(ascending=False)


violent_recid              7214
vr_charge_degree           6395
vr_case_number             6395
vr_offense_date            6395
vr_charge_desc             6395
c_arrest_date              6077
r_jail_out                 4898
r_jail_in                  4898
r_days_from_arrest         4898
r_charge_desc              3801
r_offense_date             3743
r_case_number              3743
r_charge_degree            3743
c_offense_date             1159
c_jail_out                  307
days_b_screening_arrest     307
c_jail_in                   307
out_custody                 236
in_custody                  236
c_charge_desc                29
c_days_from_compas           22
c_case_number                22
v_type_of_assessment          0
type_of_assessment            0
decile_score.1                0
v_decile_score                0
v_score_text                  0
v_screening_date              0
screening_date                0
priors_count.1                0
start                         0
end     

In [9]:
columnas_clave = [
    'sex', 'age', 'age_cat', 'race',
    'juv_fel_count', 'juv_misd_count', 'juv_other_count',
    'priors_count', 'decile_score', 'score_text',
    'two_year_recid'
]

df_clave = df[columnas_clave]
df_clave.head()


sex  age          age_cat              race  juv_fel_count  \
0  Male   69  Greater than 45             Other              0   
1  Male   34          25 - 45  African-American              0   
2  Male   24     Less than 25  African-American              0   
3  Male   23     Less than 25  African-American              0   
4  Male   43          25 - 45             Other              0   

   juv_misd_count  juv_other_count  priors_count  decile_score score_text  \
0               0                0             0             1        Low   
1               0                0             0             3        Low   
2               0                1             4             4        Low   
3               1                0             1             8       High   
4               0                0             2             1        Low   

   two_year_recid  
0               0  
1               1  
2               1  
3               0  
4               0

In [10]:


sns.countplot(x='two_year_recid', data=df_clave)
plt.title('Distribución de Reincidencia a 2 años')
plt.xlabel('¿Reincide en 2 años? (1=Sí, 0=No)')
plt.ylabel('Número de individuos')
plt.show()


NameError: name 'sns' is not defined

In [ ]:
df_two_years = pd.read_csv("/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-two-years.csv")

In [ ]:
import plotly.express as px

# Filtrado del dataset a columnas clave
columnas_clave = [
    'sex', 'age', 'age_cat', 'race',
    'juv_fel_count', 'juv_misd_count', 'juv_other_count',
    'priors_count', 'decile_score', 'score_text',
    'two_year_recid'
]
df_clave = df_two_years[columnas_clave]

# Reemplazar valores 0 y 1 por etiquetas claras en una nueva columna
df_clave['recid_label'] = df_clave['two_year_recid'].replace({0: 'No reincide', 1: 'Reincide'})
df_clave['sex_label'] = df_clave['sex'].replace({'Male': 'Hombre', 'Female': 'Mujer'})

# Diccionario para títulos personalizados
titulo_personalizado = {
    'sex_label': 'Distribución de Género',
    'race': 'Distribución de Raza',
    'age_cat': 'Distribución por Edad',
    'score_text': 'Distribución Puntaje COMPAS',
    'recid_label': 'Distribución de Reincidencia a 2 años'
}

# Mostrar gráficos
for col in ['sex_label', 'race', 'age_cat', 'score_text', 'recid_label']:
    titulo = titulo_personalizado[col]
    
    if df_clave[col].nunique() == 2:
        # Pie chart si hay solo dos valores únicos
        data = df_clave[col].value_counts().reset_index()
        data.columns = [col, 'count']
        fig = px.pie(data, values='count', names=col, title=titulo,
                     color_discrete_sequence=px.colors.qualitative.Set2)
    else:
        # Histograma para más de dos categorías
        fig = px.histogram(df_clave, x=col, color=col,
                           title=titulo,
                           category_orders={col: df_clave[col].value_counts().index.tolist()},
                           labels={col: titulo},
                           color_discrete_sequence=px.colors.qualitative.Set2)
    fig.show()

    fig.show()



/var/folders/v1/l1wd4z6509jdtvbxfqdgyp9m0000gn/T/ipykernel_22240/3557527008.py:13: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/var/folders/v1/l1wd4z6509jdtvbxfqdgyp9m0000gn/T/ipykernel_22240/3557527008.py:14: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



### Reincidencia según Puntaje COMPAS

In [ ]:
fig = px.histogram(df_clave, x="score_text", color="recid_label", barmode="group",
                   title="📊 Reincidencia según Puntaje COMPAS",
                   labels={"recid_label": "Reincidencia", "score_text": "Puntaje"},
                   category_orders={"score_text": ["Low", "Medium", "High"]},
                   color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()


### Reincidencia según raza

In [ ]:
fig = px.histogram(df_clave, x="race", color="recid_label", barmode="group",
                   title="🌍 Reincidencia según Raza",
                   labels={"recid_label": "Reincidencia", "race": "Raza"},
                   category_orders={"race": df_clave["race"].value_counts().index.tolist()},
                   color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()


### Reincidencia segun numero de antecedentes

In [ ]:
fig = px.box(df_clave, x="recid_label", y="priors_count",
             title="📦 Distribución de Antecedentes según Reincidencia",
             labels={"recid_label": "Reincidencia", "priors_count": "Número de antecedentes"},
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()


In [ ]:
# Crear gráficos adicionales para análisis bivariado con interpretación posterior

# 📌 Ejemplo 4: Reincidencia por grupo de edad
fig4 = px.histogram(df_clave, x="age_cat", color="recid_label", barmode="group",
                    title="⏳ Reincidencia según Grupo de Edad",
                    labels={"recid_label": "Reincidencia", "age_cat": "Grupo de Edad"},
                    category_orders={"age_cat": ["Less than 25", "25 - 45", "Greater than 45"]},
                    color_discrete_sequence=px.colors.qualitative.Set2)

# 📌 Ejemplo 5: Reincidencia por género
fig5 = px.histogram(df_clave, x="sex", color="recid_label", barmode="group",
                    title="🚻 Reincidencia según Género",
                    labels={"recid_label": "Reincidencia", "sex": "Género"},
                    category_orders={"sex": ["Male", "Female"]},
                    color_discrete_sequence=px.colors.qualitative.Set2)

# 📌 Ejemplo 6: Promedio de reincidencia por score (decile_score)
df_score_avg = df_clave.groupby("decile_score")["two_year_recid"].mean().reset_index()
fig6 = px.line(df_score_avg, x="decile_score", y="two_year_recid",
               title="📈 Tasa de Reincidencia Promedio según Puntaje COMPAS (decile_score)",
               labels={"two_year_recid": "Tasa promedio de reincidencia", "decile_score": "Puntaje COMPAS"},
               markers=True)

fig4.show()
fig5.show()
fig6.show()


In [ ]:
def clasificar_fallo(real, pred):
    if real == 1 and pred == 1:
        return 'True Positive'
    elif real == 0 and pred == 0:
        return 'True Negative'
    elif real == 0 and pred == 1:
        return 'False Positive'
    elif real == 1 and pred == 0:
        return 'False Negative'

df_clave['compas_outcome'] = df_clave.apply(lambda row: clasificar_fallo(row['two_year_recid'], row['compas_pred']), axis=1)


/var/folders/v1/l1wd4z6509jdtvbxfqdgyp9m0000gn/T/ipykernel_22240/475829069.py:11: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:


fig = px.bar(tabla_errores, 
             x="race", 
             y="count", 
             color="compas_outcome", 
             facet_col="sex", 
             title="Errores de COMPAS por Raza y Género",
             labels={"count": "Cantidad de casos", "race": "Raza", "compas_outcome": "Resultado del modelo"},
             color_discrete_sequence=px.colors.qualitative.Set2)

fig.show()


GLOSARIO DE EVALUACIÓN DE MODELOS
Resultado real	Predicción del modelo	Resultado	¿Qué significa?

Reincide (1)	Reincide (1)	✅ True Positive (TP)	El modelo acertó: predijo reincidencia y ocurrió.

No reincide (0)	No reincide (0)	✅ True Negative (TN)	El modelo acertó: predijo que no reincidía y no reincidió.

No reincide (0)	Reincide (1)	❌ False Positive (FP)	El modelo se equivocó: dijo que reincidía pero no fue así.

Reincide (1)	No reincide (0)	❌ False Negative (FN)	El modelo se equivocó: dijo que no reincidía pero sí lo hizo.

In [ ]:
# Total por raza
total_por_raza = df_clave.groupby('race').size().reset_index(name='total')

# Falsos positivos
falsos_positivos = df_clave[df_clave['compas_outcome'] == 'False Positive']
fp_por_raza = falsos_positivos.groupby('race').size().reset_index(name='false_positives')

# Combinar y calcular porcentaje
porcentaje_fp = pd.merge(fp_por_raza, total_por_raza, on='race')
porcentaje_fp['porcentaje_fp'] = (porcentaje_fp['false_positives'] / porcentaje_fp['total']) * 100
porcentaje_fp = porcentaje_fp.sort_values(by='porcentaje_fp', ascending=False)

# Mostrar
porcentaje_fp


race  false_positives  total  porcentaje_fp
0  African-American              805   3696      21.780303
4   Native American                3     18      16.666667
2         Caucasian              349   2454      14.221679
3          Hispanic               87    637      13.657771
5             Other               36    377       9.549072
1             Asian                2     32       6.250000

In [ ]:
# Calcular cantidad de falsos negativos por raza
falsos_negativos = df_clave[df_clave['compas_outcome'] == 'False Negative']
fn_por_raza = falsos_negativos.groupby('race').size().reset_index(name='false_negatives')

# Unir con total por raza
porcentaje_fn = pd.merge(fn_por_raza, total_por_raza, on='race')
porcentaje_fn['porcentaje_fn'] = (porcentaje_fn['false_negatives'] / porcentaje_fn['total']) * 100
porcentaje_fn = porcentaje_fn.sort_values(by='porcentaje_fn', ascending=False)

# Gráfico de barras
fig = px.bar(porcentaje_fn, x='race', y='porcentaje_fn',
             title='❗ Porcentaje de Falsos Negativos por Raza',
             labels={'race': 'Raza', 'porcentaje_fn': 'Porcentaje de Falsos Negativos'},
             color='race',
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()

# Crear gráficos individuales por tipo de resultado
# 1. Verdaderos Positivos por raza
fig_tp = px.bar(porcentajes, x='race', y='porcentaje_tp',
                title='✅ Porcentaje de Verdaderos Positivos por Raza',
                labels={'race': 'Raza', 'porcentaje_tp': 'Porcentaje de Verdaderos Positivos'},
                color='race',
                color_discrete_sequence=px.colors.qualitative.Set2)

# 2. Falsos Positivos por raza
fig_fp = px.bar(porcentajes, x='race', y='porcentaje_fp',
                title='❌ Porcentaje de Falsos Positivos por Raza',
                labels={'race': 'Raza', 'porcentaje_fp': 'Porcentaje de Falsos Positivos'},
                color='race',
                color_discrete_sequence=px.colors.qualitative.Set2)

# 3. Verdaderos Negativos por raza
fig_tn = px.bar(porcentajes, x='race', y='porcentaje_tn',
                title='✅ Porcentaje de Verdaderos Negativos por Raza',
                labels={'race': 'Raza', 'porcentaje_tn': 'Porcentaje de Verdaderos Negativos'},
                color='race',
                color_discrete_sequence=px.colors.qualitative.Set2)

fig_tp.show()
fig_fp.show()
fig_tn.show()



In [ ]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# Crear la figura con 3 filas
fig = make_subplots(rows=3, cols=1, 
                    subplot_titles=("✅ Verdaderos Positivos por Raza", 
                                    "❌ Falsos Positivos por Raza", 
                                    "✅ Verdaderos Negativos por Raza"))

# Añadir barra 1: Verdaderos positivos
fig.add_trace(go.Bar(x=porcentajes['race'], y=porcentajes['porcentaje_tp'], name="Verdaderos Positivos",
                     marker_color='seagreen'), row=1, col=1)

# Añadir barra 2: Falsos positivos
fig.add_trace(go.Bar(x=porcentajes['race'], y=porcentajes['porcentaje_fp'], name="Falsos Positivos",
                     marker_color='indianred'), row=2, col=1)

# Añadir barra 3: Verdaderos negativos
fig.add_trace(go.Bar(x=porcentajes['race'], y=porcentajes['porcentaje_tn'], name="Verdaderos Negativos",
                     marker_color='steelblue'), row=3, col=1)

# Ajustar diseño
fig.update_layout(height=900, title_text="📊 Resultados de COMPAS por Raza", showlegend=False)
fig.update_xaxes(title_text="Raza", row=3, col=1)
fig.update_yaxes(title_text="Porcentaje (%)")

fig.show()


In [ ]:


# Promedio por raza
promedio_race = df_clave.groupby('race')['decile_score'].mean().reset_index(name='Puntaje Promedio')
fig1 = px.bar(promedio_race, x='race', y='Puntaje Promedio',
              title='🔹 Promedio de Puntaje COMPAS por Raza',
              color='race',
              color_discrete_sequence=px.colors.qualitative.Set2)
fig1.show()

# Promedio por género
promedio_sex = df_clave.groupby('sex')['decile_score'].mean().reset_index(name='Puntaje Promedio')
fig2 = px.bar(promedio_sex, x='sex', y='Puntaje Promedio',
              title='🔹 Promedio de Puntaje COMPAS por Género',
              color='sex',
              color_discrete_sequence=px.colors.qualitative.Set2)
fig2.show()


## 🧠 Resumen del EDA Inicial – COMPAS

### 🔍 Distribuciones generales
- La mayoría de los casos en el dataset son de hombres (~80%) y personas African-American o Caucasian.
- Alrededor del 45% de los acusados reincidieron en los dos años siguientes al arresto.

### 📊 Reincidencia y sesgos observados
- African-Americans tienen el porcentaje más alto de **falsos positivos** (21.8%), lo que sugiere que COMPAS tiende a **sobreestimar su riesgo**.
- En cambio, grupos como Caucasian y Asian tienen más **verdaderos negativos**, lo que indica que COMPAS **predice mejor cuándo no reinciden**.
- Las mujeres parecen recibir **puntajes COMPAS ligeramente más bajos** que los hombres en promedio.

### 🔹 Puntaje COMPAS medio
- El puntaje promedio de `decile_score` es **más alto en African-Americans**, seguido de Hispanic y Native American.
- El promedio más bajo se encuentra en el grupo Asian y en personas clasificadas como "Other".

### ⚠️ Posibles hipótesis
- COMPAS podría estar influenciado por **factores raciales y de género**, ya que los puntajes no siempre reflejan la reincidencia real.
- Las diferencias en los falsos positivos podrían tener implicaciones serias en decisiones judiciales.

---

Este análisis inicial sugiere que, aunque COMPAS intenta predecir la reincidencia, puede hacerlo de forma **desigual entre grupos**, lo que requiere un análisis más profundo y el desarrollo de un modelo más justo.


In [11]:


# Cargar el dataset raw
df_raw = pd.read_csv("/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-raw.csv")

# Mostrar dimensiones, tipos de datos y primeras filas
shape_raw = df_raw.shape
dtypes_raw = df_raw.dtypes
head_raw = df_raw.head()
nulls_raw = df_raw.isnull().sum().sort_values(ascending=False)

shape_raw, dtypes_raw, nulls_raw, head_raw


((60843, 28),
 Person_ID                    int64
 AssessmentID                 int64
 Case_ID                      int64
 Agency_Text                 object
 LastName                    object
 FirstName                   object
 MiddleName                  object
 Sex_Code_Text               object
 Ethnic_Code_Text            object
 DateOfBirth                 object
 ScaleSet_ID                  int64
 ScaleSet                    object
 AssessmentReason            object
 Language                    object
 LegalStatus                 object
 CustodyStatus               object
 MaritalStatus               object
 Screening_Date              object
 RecSupervisionLevel          int64
 RecSupervisionLevelText     object
 Scale_ID                     int64
 DisplayText                 object
 RawScore                   float64
 DecileScore                  int64
 ScoreText                   object
 AssessmentType              object
 IsCompleted                  int64
 IsDeleted    

In [ ]:
df_raw[df_raw['DisplayText'] == 'Risk of Recidivism']


Person_ID  AssessmentID  Case_ID Agency_Text     LastName  FirstName  \
1          50844         57167    51950    PRETRIAL       Fisher      Kevin   
4          50848         57174    51956    PRETRIAL      KENDALL      KEVIN   
7          50855         57181    51963    PRETRIAL        DAYES     DANIEL   
10         50850         57176    51958    PRETRIAL         Debe   Mikerlie   
13         50839         57162    51945    PRETRIAL     McLaurin  Stephanie   
...          ...           ...      ...         ...          ...        ...   
60829      68595         79656    72031    PRETRIAL         PENA    ROLANDO   
60832      19968         39866    36500    PRETRIAL  BUTTERFIELD      JAMES   
60835      68598         79660    72035    PRETRIAL       SUAREZ   ANDERSON   
60838      65667         79665    72038   Probation         West      James   
60841      68603         79669    72042    PRETRIAL         Ryan    Michael   

      MiddleName Sex_Code_Text  Ethnic_Code_Text DateOfBirth  ...  \
1            NaN          Male         Caucasian    12/05/92  ...   
4            NaN          Male         Caucasian    09/16/84  ...   
7            NaN          Male  African-American    08/25/94  ...   
10        George        Female  African-American    10/09/94  ...   
13        Nicole        Female  African-American    06/29/85  ...   
...          ...           ...               ...         ...  ...   
60829          N          Male          Hispanic    06/13/85  ...   
60832    MICHAEL          Male         Caucasian    09/24/73  ...   
60835        NaN          Male         Caucasian    08/10/81  ...   
60838        NaN          Male  African-American    06/21/64  ...   
60841        NaN          Male         Caucasian    02/06/85  ...   

       RecSupervisionLevel             RecSupervisionLevelText Scale_ID  \
1                        1                                 Low        8   
4                        1                                 Low        8   
7                        4                                High        8   
10                       2                              Medium        8   
13                       1                                 Low        8   
...                    ...                                 ...      ...   
60829                    2                              Medium        8   
60832                    1                                 Low        8   
60835                    1                                 Low        8   
60838                    3  Medium with Override Consideration        8   
60841                    1                                 Low        8   

              DisplayText RawScore DecileScore ScoreText AssessmentType  \
1      Risk of Recidivism    -1.06           2       Low            New   
4      Risk of Recidivism    -1.50           1       Low            New   
7      Risk of Recidivism     0.18           8      High            New   
10     Risk of Recidivism    -0.53           6    Medium            New   
13     Risk of Recidivism    -1.55           2       Low            New   
...                   ...      ...         ...       ...            ...   
60829  Risk of Recidivism    -0.42           5    Medium            New   
60832  Risk of Recidivism    -0.36           5    Medium            New   
60835  Risk of Recidivism    -1.57           1       Low            New   
60838  Risk of Recidivism     0.16           8      High            New   
60841  Risk of Recidivism    -0.34           5    Medium            New   

       IsCompleted IsDeleted  
1                1         0  
4                1         0  
7                1         0  
10               1         0  
13               1         0  
...            ...       ...  
60829            1         0  
60832            1         0  
60835            1         0  
60838            1         0  
60841            1         0  

[20281 rows x 28 columns]

In [ ]:
# Verificar valores únicos por columna para identificar posibles problemas o duplicados
valores_unicos = df_raw.nunique().sort_values()

# Verificar valores únicos en columnas clave
columnas_clave = ['Person_ID', 'AssessmentID', 'Case_ID', 'AssessmentType', 'DisplayText',
                  'DecileScore', 'ScoreText', 'Sex_Code_Text', 'Ethnic_Code_Text', 'IsCompleted', 'IsDeleted']

valores_unicos_clave = df_raw[columnas_clave].nunique().sort_values()

# Verificar combinaciones únicas de Person_ID y DisplayText
combinaciones = df_raw.groupby(['Person_ID', 'DisplayText']).size().reset_index(name='repeticiones')

valores_unicos, valores_unicos_clave, combinaciones.head(10)


(IsDeleted                      1
 IsCompleted                    1
 AssessmentReason               1
 AssessmentType                 2
 ScaleSet                       2
 ScaleSet_ID                    2
 Sex_Code_Text                  2
 Language                       2
 ScoreText                      3
 DisplayText                    3
 Scale_ID                       3
 Agency_Text                    4
 RecSupervisionLevel            4
 RecSupervisionLevelText        4
 CustodyStatus                  6
 LegalStatus                    7
 MaritalStatus                  7
 Ethnic_Code_Text               9
 DecileScore                   11
 RawScore                     659
 Screening_Date               705
 MiddleName                  1870
 FirstName                   7225
 DateOfBirth                10382
 LastName                   10896
 Person_ID                  18610
 Case_ID                    19530
 AssessmentID               20281
 dtype: int64,
 IsCompleted             1
 IsDel

In [ ]:
# Paso 1: Filtrar solo las evaluaciones de tipo "Risk of Recidivism"
df_reincidencia = df_raw[df_raw['DisplayText'] == 'Risk of Recidivism'].copy()

# Paso 2: Seleccionar columnas relevantes
columnas_utiles = [
    'Person_ID', 'Sex_Code_Text', 'Ethnic_Code_Text', 'DateOfBirth',
    'DecileScore', 'ScoreText', 'AssessmentType', 'AssessmentReason'
]
df_reincidencia = df_reincidencia[columnas_utiles]

# Verificar datos limpios
df_reincidencia.head()

Person_ID Sex_Code_Text  Ethnic_Code_Text DateOfBirth  DecileScore  \
1       50844          Male         Caucasian    12/05/92            2   
4       50848          Male         Caucasian    09/16/84            1   
7       50855          Male  African-American    08/25/94            8   
10      50850        Female  African-American    10/09/94            6   
13      50839        Female  African-American    06/29/85            2   

   ScoreText AssessmentType AssessmentReason  
1        Low            New           Intake  
4        Low            New           Intake  
7       High            New           Intake  
10    Medium            New           Intake  
13       Low            New           Intake

In [ ]:
# Normalizar nombres en la columna de raza
df_reincidencia['Ethnic_Code_Text'] = df_reincidencia['Ethnic_Code_Text'].replace({
    'African-Am': 'African-American'
})

# Verificar valores únicos actualizados
valores_actualizados = df_reincidencia['Ethnic_Code_Text'].value_counts()

valores_actualizados


Ethnic_Code_Text
African-American    9023
Caucasian           7261
Hispanic            2914
Other                864
Asian                108
Native American       73
Arabic                25
Oriental              13
Name: count, dtype: int64

🔍 ¿Qué muestra el DecileScore de COMPAS?

El DecileScore es el puntaje de riesgo que asigna COMPAS a una persona evaluada, y va del 1 al 10, donde:

    1 = riesgo muy bajo de reincidencia

    10 = riesgo muy alto de reincidencia

Este score se calcula en base a un cuestionario automatizado, considerando:

    Edad

    Número de delitos previos

    Historial de faltas

    No necesariamente el delito actual, ni la raza (aunque el resultado puede estar correlacionado con la raza).

In [ ]:

# Gráfico 1: promedio de decile_score por raza (con raza normalizada)
promedio_score_raza = df_reincidencia.groupby('Ethnic_Code_Text')['DecileScore'].mean().reset_index(name='Puntaje Promedio')
fig_race = px.bar(promedio_score_raza, x='Ethnic_Code_Text', y='Puntaje Promedio',
                  title='🔹 Promedio de Puntaje COMPAS por Raza (Raw Dataset)',
                  labels={'Ethnic_Code_Text': 'Raza', 'Puntaje Promedio': 'Puntaje COMPAS'},
                  color='Ethnic_Code_Text',
                  color_discrete_sequence=px.colors.qualitative.Set2)

# Gráfico 2: promedio de decile_score por género
promedio_score_genero = df_reincidencia.groupby('Sex_Code_Text')['DecileScore'].mean().reset_index(name='Puntaje Promedio')
fig_sex = px.bar(promedio_score_genero, x='Sex_Code_Text', y='Puntaje Promedio',
                 title='🔹 Promedio de Puntaje COMPAS por Género (Raw Dataset)',
                 labels={'Sex_Code_Text': 'Género', 'Puntaje Promedio': 'Puntaje COMPAS'},
                 color='Sex_Code_Text',
                 color_discrete_sequence=px.colors.qualitative.Set2)

fig_race.show()
fig_sex.show()


In [ ]:
import plotly.express as px

# Calcular promedio por género
promedio_score_genero = df_reincidencia.groupby('Sex_Code_Text')['DecileScore'].mean().reset_index(name='Puntaje Promedio')

# Gráfico de pastel
fig = px.pie(promedio_score_genero, values='Puntaje Promedio', names='Sex_Code_Text',
             title='🥧 Promedio de Puntaje COMPAS por Género (Raw Dataset)',
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()


In [ ]:
# Contar cantidad de personas por género
conteo_genero = df_reincidencia['Sex_Code_Text'].value_counts().reset_index()
conteo_genero.columns = ['Género', 'Cantidad']

# Gráfico de pastel
fig = px.pie(conteo_genero, values='Cantidad', names='Género',
             title='🥧 Distribución de Personas por Género (Evaluaciones de Reincidencia)',
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()

In [ ]:

# Contar cantidad de personas por raza
conteo_raza = df_reincidencia['Ethnic_Code_Text'].value_counts().reset_index()
conteo_raza.columns = ['Raza', 'Cantidad']

# Gráfico de barras
fig = px.bar(conteo_raza, x='Raza', y='Cantidad',
             title='📊 Distribución de Personas por Raza (Evaluaciones de Reincidencia)',
             color='Raza',
             text='Cantidad',
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.update_traces(textposition='outside')
fig.show()

### 🧠 Insight - Distribución por Raza

La mayoría de las evaluaciones de reincidencia en el dataset corresponden a personas African-American y Caucasian. Esto implica que cualquier análisis sobre sesgos debe tener en cuenta el fuerte desbalance racial presente en los datos, ya que algunos grupos están poco representados (como Arabic o Oriental), lo que puede distorsionar conclusiones si no se agrupan o ajustan.


In [ ]:


# Contar cantidad de personas por nivel de riesgo (ScoreText)
conteo_scoretext = df_reincidencia['ScoreText'].value_counts().reset_index()
conteo_scoretext.columns = ['ScoreText', 'Cantidad']

# Gráfico de pastel
fig = px.pie(conteo_scoretext, values='Cantidad', names='ScoreText',
             title='🥧 Distribución de Evaluaciones por Nivel de Riesgo (ScoreText)',
             color_discrete_sequence=px.colors.qualitative.Set2)
fig.show()


### 🧠 Insight - Distribución de ScoreText

La mayoría de las personas evaluadas por COMPAS son clasificadas como de bajo o mediano riesgo. Esto sugiere que el sistema tiende a asignar scores conservadores en su evaluación general, lo que puede afectar cómo se interpretan los falsos positivos o negativos, especialmente en los extremos del score (High). Es clave tener en cuenta esta distribución para no sobreinterpretar resultados en clases poco representadas.


In [ ]:
# Filtrar filas sin valores nulos en raza o score
df_filtrado = df_reincidencia.dropna(subset=['Ethnic_Code_Text', 'ScoreText'])

# Gráfico de barras agrupadas
fig_score_raza = px.histogram(df_filtrado, x='Ethnic_Code_Text', color='ScoreText',
                               barmode='group',
                               title='📊 Distribución de Nivel de Riesgo (ScoreText) por Raza',
                               labels={'Ethnic_Code_Text': 'Raza', 'ScoreText': 'Nivel de Riesgo'},
                               color_discrete_sequence=px.colors.qualitative.Set2)
fig_score_raza.show()

### 🧠 Insight - ScoreText por Raza

Al comparar los niveles de riesgo asignados por COMPAS según la raza, se observa que el grupo African-American tiene una mayor proporción de evaluaciones clasificadas como "Medium" y "High", en comparación con otros grupos como Caucasian o Asian. Este patrón sugiere que COMPAS podría estar asignando **puntajes de riesgo desproporcionadamente altos** a ciertos grupos raciales, lo cual coincide con hallazgos de investigaciones previas y puede implicar un sesgo algorítmico importante.


In [ ]:
df_filtrado_genero = df_reincidencia.dropna(subset=['Sex_Code_Text', 'ScoreText'])

# Gráfico de barras agrupadas por género
fig_score_genero = px.histogram(df_filtrado_genero, x='Sex_Code_Text', color='ScoreText',
                                 barmode='group',
                                 title='📊 Distribución de Nivel de Riesgo (ScoreText) por Género',
                                 labels={'Sex_Code_Text': 'Género', 'ScoreText': 'Nivel de Riesgo'},
                                 color_discrete_sequence=px.colors.qualitative.Set2)
fig_score_genero.show()

### 🧠 Insight - ScoreText por Género

El análisis revela que los hombres tienden a recibir una mayor proporción de puntuaciones "Medium" y "High" en comparación con las mujeres, quienes aparecen más frecuentemente en la categoría "Low". Este patrón podría estar reflejando diferencias reales en antecedentes o simplemente una **tendencia del sistema COMPAS a asociar mayor riesgo con el género masculino**, lo que requiere un análisis más profundo para descartar sesgos injustificados.


## 🧠 Resumen del Análisis Exploratorio Inicial – COMPAS

Este análisis exploró cómo el sistema COMPAS asigna scores de riesgo de reincidencia y si lo hace de forma equitativa entre distintos grupos demográficos. Se utilizaron dos datasets: uno reducido con seguimiento de reincidencia a 2 años, y otro completo con más de 60.000 evaluaciones.

### ✅ Principales acciones realizadas:

- Limpieza y filtrado de datos para quedarnos con evaluaciones de reincidencia.
- Análisis de distribuciones generales por raza, género y nivel de riesgo (`ScoreText`).
- Cálculo de errores del sistema (TP, TN, FP, FN) por grupo racial.
- Promedio de puntajes (`DecileScore`) por raza y por género.
- Comparaciones visuales claras con gráficos de barras y pie charts (solo cuando aplicaba).

### 📊 Hallazgos clave:

- La mayoría de las personas evaluadas son hombres y African-American o Caucasian.
- African-Americans reciben una proporción significativamente mayor de scores "High" y falsos positivos.
- Los hombres presentan promedios de riesgo más altos que las mujeres.
- COMPAS tiende a clasificar la mayoría de personas en "Low" o "Medium", pero distribuye esos niveles de forma desigual.

### ⚠️ Conclusión:

Los patrones observados sugieren que el sistema COMPAS podría estar introduciendo **sesgos de riesgo** según raza y género. Estos resultados justifican la necesidad de construir un modelo alternativo más justo, que evite decisiones automatizadas potencialmente discriminatorias.


In [ ]:
conteo_riesgo_alto = df_raw_uploaded[df_raw_uploaded["ScoreText"] == "High"].shape[0]
conteo_riesgo_alto


6823

In [12]:
df_raw = pd.read_csv('/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-raw.csv')
df_two_years = pd.read_csv('/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-two-years.csv')
# Seleccionar columnas clave
df_raw_dates = df_raw[['Person_ID', 'compas_screening_date']]
df_two_years_dates = df_two_years[['id', 'screening_date']]
# Renombrar 'id' para poder hacer merge
df_two_years_dates = df_two_years_dates.rename(columns={'id': 'Person_ID'})
# Unir los dos DataFrames por 'Person_ID'
df_merged = pd.merge(df_raw_dates, df_two_years_dates, on='Person_ID', how='inner')
# Comprobar si las fechas coinciden
df_merged['fechas_iguales'] = df_merged['compas_screening_date'] == df_merged['screening_date']
# Mostrar resultado general
print(f"Fechas iguales en {df_merged['fechas_iguales'].sum()} de {len(df_merged)} registros ({(df_merged['fechas_iguales'].mean() * 100):.2f}%)")
# Si quieres ver ejemplos donde no coinciden:
print("\nEjemplos donde las fechas NO coinciden:")
print(df_merged.loc[~df_merged['fechas_iguales']].head())

KeyError: "['compas_screening_date'] not in index"

In [13]:
print(df_raw.columns)


Index(['Person_ID', 'AssessmentID', 'Case_ID', 'Agency_Text', 'LastName',
       'FirstName', 'MiddleName', 'Sex_Code_Text', 'Ethnic_Code_Text',
       'DateOfBirth', 'ScaleSet_ID', 'ScaleSet', 'AssessmentReason',
       'Language', 'LegalStatus', 'CustodyStatus', 'MaritalStatus',
       'Screening_Date', 'RecSupervisionLevel', 'RecSupervisionLevelText',
       'Scale_ID', 'DisplayText', 'RawScore', 'DecileScore', 'ScoreText',
       'AssessmentType', 'IsCompleted', 'IsDeleted'],
      dtype='object')


In [ ]:
# Cargar el dataset
df = pd.read_csv('/Users/danielalejandroalvarez/Documents/Master Nuclio Python/Entregable FInal/compas-scores-two-years.csv')
# Comprobar si son iguales
df['fechas_iguales'] = df['compas_screening_date'] == df['screening_date']
# Mostrar resultado general
print(f"Fechas iguales en {df['fechas_iguales'].sum()} de {len(df)} registros ({(df['fechas_iguales'].mean() * 100):.2f}%)")
# Mostrar ejemplos donde no coinciden
print("\nEjemplos donde las fechas NO coinciden:")
print(df.loc[~df['fechas_iguales'], ['compas_screening_date', 'screening_date']].head())